### PR TITLE
BREAKING: rename `http.client_ip` to `http.forwarded.for` and introduce `http.forwarded.proto` attributes

### LINKED ISSUE #2338: Is it ok for http.scheme to factor in forwarded "proto" field / x-forwarded-proto headers?
**What are you trying to achieve?**

Reduce user confusion/panic about why we report a server request as `http` when it was "really" `https`.

This happens commonly when terminating SSL in front of the process that is capturing telemetry.

I'd like to respect the forwarded/x-forwarded-proto headers when capturing `http.scheme`.

But if that is not acceptable, it would be helpful to introduce a new attribute to capture the forwarded "proto" field / x-forwarded field so that backends can decide to display that instead if they prefer.

### PR DESCRIPTION
**Pull request by lmolkova**
Fixes #2338

`http.client_ip` attribute does not work well with new(ish) [forwarded](https://www.rfc-editor.org/rfc/rfc7239.htm) header definition:
- in the RFC, [`for` component](https://www.rfc-editor.org/rfc/rfc7239.html#section-5.2) is a list of nodes, which can be IP addresses, IP:port combinations, or an obfuscated identifiers. This makes `client_ip` not precise as a node is not necessarily an IP.
- there is a desire to add other forwarded components: `proto` or `host`. Having a descriptive namespace to populate all of them is preferable. Adding `http.client_proto` and `http.client_host` could be confusing.

## Changes

- [BREAKING] renames `http.client_ip` to `http.forwarded.for`
- introduces `http.forwarded.proto`

Full `forwarded` value can be captured by enabling collection of the corresponding header (`http.request.header.forwarded`), same goes for `x-forwarded-*` headers.

**Additional context:**

- ECS supports `network.forwarded_ip`, but it has similar limitations as `http.client_ip` see discussions in ECS: https://github.com/elastic/ecs/issues/938
- while reverse proxies support other-than-HTTP protocols, I was not able to find any analogs to forwarded headers. Given that header is specified for HTTP only, it makes sense to put it into `http`  namespace.

**Alternative attribute name:**

since `for` component in `forwarded` header is not a single address, but a list of them, it could be better to call the corresponding attribute `http.forwarded.addr`

### CODE REVIEW

File: semantic_conventions/trace/http.yaml [hunk 1]
```diff
@@ -99,24 +99,38 @@ groups:
...
+      - id: forwarded.proto
+        type:
+          allow_custom_values: true
+          members:
+            - id: http
+              value: "http"
+            - id: https
+              value: "https"
+        requirement_level:
+          recommended: if known and only if different from `http.scheme`
```
Comment (trask) [inline semantic_conventions/trace/http.yaml:129]: to match definition of `forwarded.for`?
```suggestion
          recommended: if and only if it's available.
```

File: semantic_conventions/trace/http.yaml [hunk 2]
```diff
@@ -99,24 +99,38 @@ groups:
...
+        brief: >
+          The original protocol client used to connect to the proxy (e.g.
+          from the `proto` component of [Forwarded](https://www.rfc-editor.org/rfc/rfc7239.html) header)
+        examples: ['https', 'ftp']
```
Comment (trask) [inline semantic_conventions/trace/http.yaml:133]: maybe, as more common values?
```suggestion
        examples: ['https', 'http']
```

**Review by trask — APPROVED** [review]

**Review by jsuereth — APPROVED** [review]

File: semantic_conventions/trace/http.yaml [hunk 3]
```diff
@@ -99,24 +99,38 @@ groups:
         requirement_level: required
         sampling_relevant: true
         examples: ['/path/12314/?q=ddds']
-      - id: client_ip
+      - id: forwarded.for
```
Comment (MSNev) [inline semantic_conventions/trace/http.yaml:102]: While I support adding an additional `http.fowarded.xxx` namespace, I'm not supportive of renaming the existing `client_ip` as not all use-cases will be forwarded as not all systems are behind Proxies or transparent proxy environments.

Comment (lmolkova) [reply semantic_conventions/trace/http.yaml:102]: if something is not behind a proxy, it MUST NOT use `http.client_ip`  even according to the current spec - it should use `net.peer.name` - this is not changing.

Comment (MSNev) [reply semantic_conventions/trace/http.yaml:102]: Well, that's not the way I read it, there is no `MUST NOT`, it says it [should be set if different](https://github.com/open-telemetry/opentelemetry-specification/blob/0d9081669c0c087c092bf35fe5c5652c2b505b94/semantic_conventions/trace/http.yaml#L111-L118) and then goes on to say

> Setting `http.client_ip` when it's the same as `net.sock.peer.addr` means that one is at least somewhat confident that the address is not that of the closest proxy.

So anyone could quite easily interrupt this as this "IS" the real client IP... (Although from a client side perspective, this is PII and I would also want the scrubbed in some manner)

Comment (MSNev) [reply semantic_conventions/trace/http.yaml:102]: And looking at the JS repo (for Node) the (experimental) http instrumentation, is setting this from the `X-FORWARDED-FOR` header (if available). So it's very likely that someone is using this.

@hectorhdzg do you know if we are using this instrumentation today? If so, be aware of this possible change.

Comment (trask) [reply semantic_conventions/trace/http.yaml:102]: > And looking at the JS repo (for Node) the (experimental) http instrumentation, is setting this from the `X-FORWARDED-FOR` header (if available).

this is how we're populating `http.client_ip` in Java HTTP instrumentation as well.

my understanding is that this is currently the only purpose of `http.client_ip` (to capture `x-forwarded-for` or `forwarded[for]` on SERVER spans).

**Review by reyang — APPROVED** [review]

File: semantic_conventions/trace/http.yaml (hunk 3 above)
Comment (lmolkova) [reply semantic_conventions/trace/http.yaml:102]: @MSNev spec language is confusing here. I read it as: if you only have peer address from socket-level info, do not set client_ip because you have no other source of information to know if it's the real one.

I assume some custom instrumentation can know it's the real one and then populate http.client_ip, but it's not the case for OTel auto-instrumetnations.

**Comment by trask** [comment]
> I doubt this is a replacement for http.client_ip. Should these attributes maybe be added in addition? There may be sources of client_ip other than these HTTP headers. (weak opinion, I'm not very sure here)

since we aren't sure of the use case for `http.client_ip`, I'd suggest that we drop it before stability, and we can always add it (or something similar) back in the future

**Comment by MSNev** [comment]
> I doubt this is a replacement for http.client_ip. Should these attributes maybe be added in addition? There may be sources of client_ip other than these HTTP headers. (weak opinion, I'm not very sure here)

Yep a quick search show lots of different ones
- X-Client-IP
- X-Forwarded-For
- CF-Connecting-IP (Cloudflare)
- Fastly-Client-IP (Fastly CDN)
- True-Client-IP  (Akamai, Cloudflare)
- X-Real-IP (Nginx / FastCGI)
- X-Cluster-Client-IP (Rackspace, Riverbed)

Node npm package [request-ip](https://www.npmjs.com/package/request-ip)

**Comment by trask** [comment]
`X-Client-Id` - I couldn't find a definition of this, so it's hard to say what the semantics of it are.

`X-Forwarded-For` is covered by this PR's `http.forwarded.for`

`CF-Connecting-IP` is "the client IP address connecting to Cloudflare to the origin web server" (which could still be a proxy itself and not the original client), so it doesn't represent `http.client_ip` anyways, which is "the IP address of the original client behind all proxies". If/when a semconv is needed for `CF-Connecting-IP`, maybe it makes more sense for it to be under `cloudflare.*` conventions.

`Fastly-Client-IP` is similarly defined as "the non-Fastly thing that is making the request to Fastly", which also doesn't match `http.client_ip`. If/when a semconv is needed for `Fastly-Client-IP`, maybe it makes more sense for it to be under `fastly.*` conventions. (or maybe there's some new generalization to be made between fastly and cloudflare)

`True-Client-IP` - "There is no difference between the True-Client-IP and CF-Connecting-IP headers besides the name of the header"

I didn't look up the others, but I'd still suggest that we drop `http.client_ip` for now, before stability. We can always add it back later if we discover we were wrong and there's a need for it.

**Comment by trask** [comment]
one other thought, `http.forwarded.for` in this PR is still defined as

> The address of the original client behind all proxies.

and while typically that will come from the `Forwarded` or `X-Forwarded-For` header, it's not saying it _can't_ come from another header.

**Comment by lmolkova** [comment]
the problem with `client_ip` that `forwarded` HTTP standard header does not necessarily carries an IP, it rather recommends obfuscating it. Given that `forwarded` header is part of HTTP standard, we should support it properly. 

 If we keep `client_ip` attribute it'd serve unclear and narrow scenarios.  `forwarded.for` allows to put original IP into it including any of the x-something headers (if they represent original IP).

**Comment by scheler** [comment]
Can `http.request.header.<key>` not be used for the purpose?

**Comment by lmolkova** [comment]
> Can `http.request.header.<key>` not be used for the purpose?

Not really.

here we capture individual components - `proto` (scheme) and the first address in `for` component. Also these attributes put a burden on instrumentation to support variety of `forwarded`, `x-forwarded-*` and other attributes, hiding this from users and providing attributes by default.

 `http.request.header.<key>` puts all burden on users and requires them to enable it explicitly.

File: semantic_conventions/trace/http.yaml (hunk 3 above)
Comment (lmolkova) [reply semantic_conventions/trace/http.yaml:100]: I'm resolving this thread assuming the concern is addressed, if it's not the case, @MSNev, please unresolve

File: schemas/1.19.0 [hunk 4]
```diff
@@ -4,7 +4,7 @@ versions:
   1.19.0:
     spans:
       changes:
-        # https://github.com/open-telemetry/opentelemetry-specification/pull/3209
+        # https://github.com/open-telemetry/opentelemetry-specification/pull/3209
```
Comment (trask) [inline schemas/1.19.0:7]:
```suggestion
        # https://github.com/open-telemetry/opentelemetry-specification/pull/3209
```

**Comment by trask** [comment]
@Oberon00 @MSNev @scheler have your concerns above been addressed?

**Comment by MSNev** [comment]
> @trask @Oberon00 @MSNev Nev Wylie FTE @scheler have your concerns above been addressed?

Simplistically No, but I'm not going to the extreme and block the PR.

My main concerns are that people are / may already using this value and based on the existing comments it is "suggested" that this is the closest (single) field which identifies the actual client ip (as far as can be determined). While the new wording explicitly prohibits setting this value if nothing exists. Meaning there will now be no single location for anyone to "check". 

There is also an implicit assumption introduced by renaming this to "forward.for", the user is always behind some form of intervening proxy (which will not "always" be the case), which is reinforced by the "SHOULD NOT" set. So this (forces) complications for backend / queries in that now they need to always check this new field AND `net.sock.peer.addr` (because forward.for should not be set while client_ip calls out if they are the same).

Additionally, and as with a lot of the existing OpenTelemetry implementations, this reinforces the server side view (receiver) within the API and the semantic conventions, ignoring how any existing usages for "senders" (clients, who "know" their own IP address (public or not) -- eg. in-house private network monitoring), this use case does not map to forward.for unless you squint and say I sent a request me (local instrumentations that wrap "sending" API's and create Span's, Logs' or Events for this).

Keeping `http.client_ip` would allow the instrumentations to provide the "real" client ip as part of the request (in these cases) as the location is put this "known" value.

> since we aren't sure of the use case for http.client_ip, I'd suggest that we drop it before stability, and we can always add it (or something similar) back in the future

So I disagree.

My preference would be to keep the `http.client_ip` as the final "container" for what is believed to be the real client ip (after resolving all sources) keeping similar wording to what already exists...

> `http.client_ip` when it's the same as `net.sock.peer.addr` means that one is at least somewhat confident that the address is not that of the closest proxy.

While still adding the `forwarded.for` to support this known use case of the forwarding headers. 

In terms of similar prior-art for this scenario ECS does provide the [forward-for](https://www.elastic.co/guide/en/ecs/current/ecs-network.html#field-network-forwarded-ip) (which is where I assume this came from, AND the client [address](https://www.elastic.co/guide/en/ecs/current/ecs-client.html#field-client-address) / [ip](https://www.elastic.co/guide/en/ecs/current/ecs-client.html#field-client-ip) (albeit in it's own general `client` namespace).

So to summarize my preference
- keep `http.client_ip`
- If/when ECS and the `client` namespace is merged, then at that point I would be supportive of a possible renaming of this field.

**Comment by lmolkova** [comment]
I hear several concerns, please correct me if I miss something:
1. we don't have an attribute to represent 'best known client address', which could be either behind proxy or a direct peer. Note that **this PR does not change status quo on this front, `http.client_ip` represented IP behind proxy before it.**
2. the client-side concern. I'm not sure how it applies to server-only `http.client_ip` attribute.
3. breaking change. It's still an experimental spec and also our last chance to fix naming. there likely will be more breaking-changes coming.

Additionally, client_ip is not a good name for something that's not necessarily an IP.

So, it's possible that we'll need to introduce other attributes that serve best-known-original-client-address in the future or rename and repurpose the `http.client_ip` aka `http.forwarded.for` to that one.

**Comment by Oberon00** [comment]
I can't really add more in this discussion. Just one thing:

> While the new wording explicitly prohibits setting this value if nothing exists.

I think client_ip already must not be set if you have no information beyond net.sock.peer.ip.

File: semantic_conventions/trace/http.yaml [hunk 5]
```diff
@@ -97,24 +97,36 @@ groups:
...
-            This attribute should be set when a source of information different
-            from the one used for `net.sock.peer.addr`, is available even if that other
-            source just confirms the same value as `net.sock.peer.addr`.
-            Rationale: For `net.sock.peer.addr`, one typically does not know if it
-            comes from a proxy, reverse proxy, or the actual client. Setting
-            `http.client_ip` when it's the same as `net.sock.peer.addr` means that
-            one is at least somewhat confident that the address is not that of
-            the closest proxy.
-        examples: '83.164.160.102'
+            This attribute SHOULD be set when a source of information different
```
Comment (trask) [inline semantic_conventions/trace/http.yaml:112]:
```suggestion
            This attribute SHOULD be set when the source of information is different
```

File: specification/trace/semantic_conventions/http.md [hunk 6]
```diff
@@ -247,17 +248,14 @@ If the route cannot be determined, the `name` attribute MUST be set as defined i
...
-
-This attribute should be set when a source of information different
-from the one used for `net.sock.peer.addr`, is available even if that other
-source just confirms the same value as `net.sock.peer.addr`.
-Rationale: For `net.sock.peer.addr`, one typically does not know if it
-comes from a proxy, reverse proxy, or the actual client. Setting
-`http.client_ip` when it's the same as `net.sock.peer.addr` means that
-one is at least somewhat confident that the address is not that of
-the closest proxy.
+This attribute SHOULD be set when a source of information different
```
Comment (MSNev) [inline specification/trace/semantic_conventions/http.md:256]: Spoke with trask offline around this to highlight the "source" nature of this description.
I would suggest adding or keeping the wording to further highlight the source (vs value) of the obtain information.

eg. I think the following in the existing clearly states that this should be set
` is available even if that other source just confirms the same value`

while I originally mis-read the updated version to indicate that the value should not be set if it's the same as `net.sock.peer.addr`

**Comment by MSNev** [comment]
The only other remaining item is the breaking nature of this change, especially now that with merging with ECS that this would likely change yet again. If this change is delayed and included in the bigger set of breaking (ECS merge) changes I think this would be better and then my concerns would be reduced.

**Comment by trask** [comment]
we discussed in HTTP semconv stability WG and agreed that this change should be bundled with the other HTTP semconv breaking changes and covered by the transition period we are defining in #3362
